x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/snapshot/upgrade/SnapshotUpgradeState.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.job.snapshot.upgrade;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import java.io.IOException;
+import java.util.Locale;
+
+public enum SnapshotUpgradeState implements Writeable {
+
+    READING_NEW_STATE, STOPPED, FAILED;
```

Comment:
I didn't opt for a "writing_old_state" or an "opened" state as neither really conveyed any information. If the state is null, we know that either it is not assigned to a node or it is assigned and still loading the old snapshot. 

Once we are in the `reading_new_state`, then that indicates that we have reached the point of no return and any failure from that state indicates a corrupted job model snapshot.

Reply:
`reading_new_state` is very much from the perspective of the Java code rather than the end user.  As an end user who doesn't even know that the code is split between Java and C++ I would have thought `writing_new_state` makes more sense.  Or `saving_new_state` would be a compromise that makes sense to both end users and Java developers.

I would also introduce a `reading_old_state` or `loading_old_state` enum value that can be used in stats and API responses instead of `null`.  We went through that cycle with job states.  Initially there was no `opening` state, because a `null` task state basically meant that.  But then we found it was nicer to have a specific enum value for it and translate `null` to that enum value in some places.  Even if it's not used anywhere it will avoid BWC code to add it to the enum from the outset.